data/appRateLimiter.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/util"
+	"golang.org/x/crypto/blake2b"
+
+	"github.com/algorand/go-deadlock"
+)
+
+const numBuckets = 128
+
+type keyType [8]byte
+
+// appRateLimiter implements a sliding window counter rate limiter for applications.
+// It is a shared map with numBuckets of maps each protected by its own mutex.
+// Bucket is selected by hashing the application index with a seed (see memhash64).
+type appRateLimiter struct {
+	maxBucketSize        uint64
+	serviceRatePerWindow uint64
+	serviceRateWindow    time.Duration
+
+	// seed for hashing application index to bucket
+	seed uint64
+	// salt for hashing application index + origin address
+	salt [16]byte
+
+	buckets [numBuckets]map[keyType]*appRateLimiterEntry
+	mus     [numBuckets]deadlock.RWMutex
+	lrus    [numBuckets]*util.List[keyType]
+
+	// evictions
+	// TODO: delete
+	evictions    uint64
+	evictionTime uint64
+}
+
+type appRateLimiterEntry struct {
+	prev       atomic.Uint64
+	cur        atomic.Uint64
+	interval   atomic.Int64 // numeric representation of the current interval value
+	lruElement *util.ListNode[keyType]
+}
+
+// makeAppRateLimiter creates a new appRateLimiter from the parameters:
+// maxCacheSize is the maximum number of entries to keep in the cache to keep it memory bounded
+// maxAppPeerRate is the maximum number of admitted apps per peer per second
+// serviceRateWindow is the service window
+func makeAppRateLimiter(maxCacheSize uint64, maxAppPeerRate uint64, serviceRateWindow time.Duration) *appRateLimiter {
+	// convert target per app rate to per window service rate
+	serviceRatePerWindow := maxAppPeerRate * uint64(serviceRateWindow/time.Second)
+	maxBucketSize := maxCacheSize / numBuckets
+	r := &appRateLimiter{
+		maxBucketSize:        maxBucketSize,
+		serviceRatePerWindow: serviceRatePerWindow,
+		serviceRateWindow:    serviceRateWindow,
+		seed:                 crypto.RandUint64(),
+	}
+	crypto.RandBytes(r.salt[:])
+
+	for i := 0; i < numBuckets; i++ {
+		r.buckets[i] = make(map[keyType]*appRateLimiterEntry)
+		r.lrus[i] = util.NewList[keyType]().AllocateFreeNodes(int(maxBucketSize))
+	}
+	return r
+}
+
+func (r *appRateLimiter) entry(b int, key keyType, curInt int64) (*appRateLimiterEntry, bool) {
+	r.mus[b].Lock()
+	defer r.mus[b].Unlock()
+
+	if len(r.buckets[b]) >= int(r.maxBucketSize) {
+		// evict the oldest entry
+		start := time.Now()
+		atomic.AddUint64(&r.evictions, 1)
+
+		el := r.lrus[b].Back()
+		delete(r.buckets[b], el.Value)
+		r.lrus[b].Remove(el)
+
+		atomic.AddUint64(&r.evictionTime, uint64(time.Since(start)))
+	}
+
+	entry, ok := r.buckets[b][key]
+	if ok {
+		el := entry.lruElement
+		r.lrus[b].MoveToFront(el)
+	} else {
+		el := r.lrus[b].PushFront(key)
+		entry = &appRateLimiterEntry{lruElement: el}
+		entry.cur.Store(1)
+		entry.interval.Store(curInt)
+		r.buckets[b][key] = entry
```

Comment:
Since the entry is created and destroyed only in this same locked method, maybe using a sync pool is an easy reach.

Reply:
sure, in my list after adding some final benchmarks to get a base line performance.